### DISCUSSION TITLE
People: Add option to run face detection at a specific picture from the UI

### DISCUSSION THREAD

**Discussion by kingvuke** [photoprism/photoprism#3783] [Q&A]
As a user, I would like to perform an face detection from the user interface.

![Screenshot 2023-09-29 204045](https://github.com/photoprism/photoprism/assets/104008018/42d6bb7e-4664-4bba-94f2-acb0141d7580)

Therefor here maybe 2 Buttons could be added:
Start face detection
Rescan face detection

**Comment by graciousgrey** [comment]
Face detection is performed automatically during indexing. If a face could not be detected during the first indexing, it will not be detected later, unless the photo was changed.

**Comment by kingvuke** [comment]
From my side, I have had a different experience. I'll try to reproduce this the days.

I add pictures to the collection. It is all indexed and the system had several days. In the latest images the face is missing completely or partially. Then when I add the next batch of images, faces are recognized before them. I do not change the pictures.

Please look at this picture, it does not recognize the faces here. These are clearly visible, and grayed out only for here ;). 
![grafik](https://github.com/photoprism/photoprism/assets/104008018/0b57c893-a8fc-4d3a-bff3-b116c316e784)

Photoprism runs on my NAS in an Docker with 16GB Ram and an AMD Ryzen 3 PRO 4350G. The System is bored.

**Reply by graciousgrey** [reply]
> From my side, I have had a different experience. I'll try to reproduce this the days.

That's interesting, we never experienced this before. Let us know if you can reproduce this.
You can use the command `docker compose exec photoprism photoprism faces index` to trigger the face scan again from a terminal, this might help you with testing.

https://docs.photoprism.app/getting-started/docker-compose/#command-line-interface

**Comment by kingvuke** [comment]
> > From my side, I have had a different experience. I'll try to reproduce this the days.
> 
> That's interesting, we never experienced this before. Let us know if you can reproduce this. You can use the command `docker compose exec photoprism photoprism faces index` to trigger the face scan again from a terminal, this might help you with testing.
> 
> https://docs.photoprism.app/getting-started/docker-compose/#command-line-interface

I had already done that. My shell script is doing:

docker exec photoprism-photoprism-1 photoprism faces index
docker exec photoprism-photoprism-1 photoprism faces update
docker exec photoprism-photoprism-1 photoprism faces optimize
docker exec photoprism-photoprism-1 photoprism faces stats


The problem here is that this runs for a very very long time. Also a reason for the buttons that I wish.

**Reply by lastzero** [reply]
If you indexed your files, face detection was enabled (!) and no faces were detected, then no faces should be detected when you repeat the process, unless the image has changed or face detection was not enabled the first time (or you ran the faces reset command). Why would running the same code a second time yield different results?

**Reply by kingvuke** [reply]
I can't tell you that. I can't prove that yet. As I said, I have the feeling that he recognizes it more later.

I don't want to put the photo in grayed out. But I think you can see that well enough. Why are there no faces recognized?
The faces are large enough, also the file itself (2592 × 1940, 2.7 MB).

**Reply by lastzero** [reply]
The face detection isn't perfect and 100% reliable. That said, it should yield the same results unless something has changed in the meantime, like the settings or the image being analyzed.

**Reply by kingvuke** [reply]
I understand. 
But what I don't understand is that the system recognizes the faces in images with poor quality, smaller resolution and smaller faces.

Exactly for such a reason a button to rescan the image would not be bad with the possibility of log outputs ;)

**Reply by lastzero** [reply]
I can't give you an answer without testing it with your original file and settings. Unfortunately, we are currently busy with other tasks. The debug outputs generated during indexing and visible in the web user interface are most likely not sufficient for debugging, especially since the log level in these logs is not detailed enough.

**Comment by dgtal1** [comment]
Did you find any clues why it happened? Were there any fixes provided in the mean time? In my case I noticed that my kid has only ca 300 photos indexed for his face. Rouhgly there should be at least 10000-15000 instead. About 5000 of his photos were taken with a DSLR camera with a really good quality. Some of these DSLR photos were recognized by Photoprism and faces were assigned, but only some of them. 

I've run the `facial index` CLI command and waiting for results. Hopefully it will pick them up and match correctly. I saw that Photoprism deetcted faces in a really much worse conditions like people somewhere in the background or low light conditions. But these photos of my kid taken with HQ camera and in good or very good ligh conditions were not detected for some reason - I mean no faces were detected. 
I feel there must be some issue with it then.

> If you indexed your files, face detection was enabled (!) and no faces were detected, then no faces should be detected when you repeat the process, unless the image has changed or face detection was not enabled the first time (or you ran the faces reset command). Why would running the same code a second time yield different results?

We don't know that eiter. We don't know/understand how photoprism is built and how does this work, but there must be some factors affecting the facial recognition that it is not done properly for some HQ samples. Maybe there is some bug somewhere? I believe the process depends on data extracted from the photos, so maybe some bug is affecting the input data and the facial recognition fails? 
I understand software is deterministic and running the same code should give the same result, but also for this reason - there must be an explanation why it works fine for a much worse samples and it doesn't for the great ones.

Do you have any finding about this since Oct 2023?

**Reply by kingvuke** [reply]
Photoprism is using Pigo for detection.
You can see that here:
https://docs.photoprism.app/developer-guide/metadata/face-recognition/

I downloaded Pigo localy from here https://github.com/esimov/pigo/releases
and the facefinder from here https://github.com/esimov/pigo/blob/master/cascade/facefinder
Then i tried with different settings with some of my pictures. But the result was bad.

I used also the picture from my post of from 3th of october. Pigo detected directly one face with default settings. But i was never able to detect 2 faces. My result was: 0, 1 or >20 fake faces. I have no idea why.

At the end i switch to immich.
Here are the recognized pictures with the faces of my daughter, my wife and me
Photoprism: 846 daughter, 2404 wife, 4206 me
Immich: 21265 daughter, 15134 wife, 9224 me

**Reply by dgtal1** [reply]
Whoa...a huge difference in numbers! I expected also a very similar kind of proportion in the numbers for my Library. Photoprism found about 3.5k thousand of my own faces. It required barely 0 attention from my side to tag my faces correctly. However for my kid it found only about 300 whereas it should find at least about 10000. After spending hours to tag my kid manually - the number raised to 2.5k. Still definitely too low.

I use Synology Moments and I'm stunned with its object&face detection accuracy - my kid scored about 9.5k matches, my face number was similar to Photoprism. To get the faces of my kid right on Moments I had to merge maybe 10-15 faces, whereas in Photoprism I had to merge/tag hundreds and it's still not even close to the right number. It is somehow not able to handle people, who changed a lot over time. Kids grow, change hair, glasses, make funny smiles. That's probably difficult to handle, but is definitely possible - see Moments - no problem with this.

I'd still use Moments, but it's abandoned and no longer developed. Also it's sluggish (due to db stored on HDD on my NAS, and no way to move it to my external SSD as I did for Photpprism) and the mobile app is crap. The web version is not responsive, so on mobile devices or tvs it looks and works crappy.

Photoprism UI is superior to Moments. I generally like the approach and features. Altough facial recognition (and object recognition) is far from perfect. It's that comopnent, that developers should focus on most to make it right. Other than that I find Photoprism a really good indexing solution: Calendar, Places, search, Albums - these things work really great.

Altough I've just checked immich and it looks very promising! It definitely tries to resemble Google Photos (timeline) and I like this approach (altough performance-wise Photoprisms/s calendar is better for slower devices with limited resouorces), but I wonder if it's gonna run well on my tv. I will definitely try it out! Thanks!

**Reply by lastzero** [reply]
Why not help improve Pigo? That would be the easiest solution, but nobody seems to consider that... Open Source doesn't mean one person works for free and everyone else consumes and points out what should be done.

**Reply by dgtal1** [reply]
To help with such thing is beyond my abilities unfortunately. I'm really impressed with Protoprism in general and I like it. It's just that face & object recognition that's underperforming. As an alternative to Pigo - maybe Photoprism should try to switch to another component that gives better results? Maybe that would be easier or the results would imporove much faster? Using external components always has one drawback - you never know when is it going to be abandoned and you never know the development pace. There are tons of open source solutions similar to each other. Sometimes it's hard to judge which horse to bet on. I fully understand this. 

Anyway guys - I will you luck and hoping Photoprism will evelove in the right direction ti address the current issues. It is devenitely a solution that deserves attention and recognition in the industry.

**Comment by dgtal1** [comment]
faces indexing complete. No new recognitions. Something's really wrong with the matching algorithms in Photoprism.

**Comment by nickbreen** [comment]
Running `faces index`... observed the log, but I can't quite grok what it's telling me. Snippet:

```
INFO[2024-08-14T07:56:28Z] index: found 1 face in 20120102_135715.jpg [217.375582ms] 
INFO[2024-08-14T07:56:28Z] index: skipped main jpg file Pictures/20120102_135715.jpg 
INFO[2024-08-14T07:56:28Z] index: skipped main jpg file Pictures/20120102_135727.jpg 
INFO[2024-08-14T07:56:28Z] index: found 1 face in 20120102_135738.jpg [246.644005ms] 
INFO[2024-08-14T07:56:28Z] index: skipped main jpg file Pictures/20120102_135738.jpg 
INFO[2024-08-14T07:56:28Z] index: skipped main jpg file Pictures/20120102_135754.jpg 
INFO[2024-08-14T07:56:28Z] index: skipped main jpg file Pictures/20120102_135757.jpg 
INFO[2024-08-14T07:56:29Z] index: skipped main jpg file Pictures/20120102_135800.jpg 
```

It is indicating that it is _skipping_ various files (e.g. `20120102_135727.jpg`). Does that mean it found no faces or that it didn't try?

It also logs that it found 1 face and then logs that it skipped that file! E.g. `20120102_135715.jpg`. Does that mean it found a face but ignored it?
